### ISSUE TITLE
Train deep architecture for trafic sign prediction

### ISSUE THREAD

**Issue by albanD** [Moodstocks/gtsrb#1]
task: Develop from end to end the code to train a deep architecture following the paper in the README and evaluating its performances.

forked repo: [albanD/gtsrb](https://github.com/albanD/gtsrb)

Main tasks to be performed:
- [x] Basic pipeline
  - Image io and cropping
  - MLP declaration
  - MLP training
  - MPL testing on validation set
- [x] IO normalization
  - Mean subtraction has been done
  - Jittering (but the scale)
- [x] Use the test set 
- [x] Use of CNN architecture
- [x] Test torch.optim SGD ( other optimization functions?)
- [x] Use batch of size > 1
  - Not possible with nn.StochasticGradient which can only use batch size of 1
  - Will be done with the torch.optim version of the optimization
- [x] Multi-scale training
- [x] Use feature extractor in matching mode

**Comment by albanD** [comment]
Results:
1. The basic MLP pipeline gives 8.91% error (using commits up to 9b152a9658ea93c963744cad9206f7577fb11af0).
   
   Runtime (dataset reading + training (25 epoch) + testing):
   real  12m9.548s
   user  17m34.548s
   sys 1m15.485s
2. The basic CNN gives 6.23% error (using commits up to e04f42d54c55adda9f74db71485a18034d34f1c0)
   
   Runtime (dataset reading + training (10 epoch) + testing):
   real  24m33.187s
   user  31m51.297s
   sys 3m36.943s
3. The basic CNN using torch.optim SGD with batch size of 1 gives 7.5% error (up to 24918727544be4a8c00fa758d1663940e4705932)
   
   Runtime (dataset reading + training (10 epoch) + testing):
   (Other computations where running at the same time were slowing it down)
   real  40m15.428s
   user  47m46.617s
   sys 2m26.566s
4. The basic CNN using torch.optim SGD with batch size of 100 gives 34.22% error (not converged yet)
   
   Runtime(pre-stored dataset + training(10 epoch) + testing):
   real  23m42.640s
   user  33m31.560s
   sys 2m17.919s
5. The basic CNN using torch.optim CG with batch size of 100 gives 10.13% error
   
   Runtime(pre-stored dataset + training(2 epoch) + testing):
   real  14m7.321s
   user  20m17.809s
   sys 1m20.432s

**Comment by albanD** [comment]
Final test using multi-scale network but no jittering:
- Using CG:
  - Maximun number of iterations set to 2:
    - 50% error on validation set after first epoch
    - 12% error on validation set after second epoch
    - 10% error on validation set after third epoch
    - after 3 epoch, 8.6% error on test set.
  - Maximun number of iterations set to 5:
    - 21% error on validation set for first epoch
    - 18% error on validation set for second epoch
    - oscillate around 20% for next epochs
    - after 30 epoch, 24% precision on test set
    - matching mode gives 66% error rate
- Using SGD
  - Takes 7 epoch to get under 10% error rate on validation set
  - after 100 epochs, 4.65% error rate on validation set
  - final result of 6.91% on test set
  - matching mode gives 68% error rate
